File: scratch/Luhn_Algorithm/Luhn.scala
def luhn(creditCard: String): Boolean = {
  val (odds, evens) = creditCard.reverse.zip(Range(1, creditCard.size + 1)).partition { case ((ch, n)) => n % 2 != 0 }
  val s1 = odds.map { case ((ch, _)) => ch.asDigit }.sum
  val s2 = evens.map { case ((ch, _)) => {     
    val doubled = ch.asDigit * 2 
    if (doubled < 10) doubled else doubled % 10 + 1
  }}.sum
  (s1 + s2) toString() endsWith("0")  
}

val (validNumbers, invalidNumbers) =
    (List("2621195162335", "49927398716", "1234567812345670", "4485284720134093"),
      List("49927398717", "1234567812345678"))

val creditCards = validNumbers ++ invalidNumbers
println(creditCards filter luhn)

File: scratch/Luhn_Algorithm/luhn.hs
import Data.List
import Data.Char

sumOfDigits n = if n > 9 then (n `mod` 10) + 1 else n

luhn :: String -> Bool
luhn creditCard = 
  "0" `isSuffixOf` show (s1 + s2)
  where
    (odds, evens) = partition (\(ch,n) -> n `mod` 2 == 1) $ zip (reverse $ map digitToInt creditCard) [1..]
    s1 = sum $ fst $ unzip odds
    s2 = sum $ map sumOfDigits $ map (*2) $ fst $ unzip evens
  
main :: IO ()
main = do
  let creditCards = ["2621195162335", "49927398716", "1234567812345670", "4485284720134093"] ++ ["49927398717", "1234567812345678"]
  print $ filter luhn creditCards
